moveit_ros/hybrid_planning/local_planner/local_constraint_solver_plugins/src/servo_solver.cpp

```diff
@@ -0,0 +1,174 @@
+  return true;
+}
+
+bool ServoSolver::reset()
+{
+  RCLCPP_INFO(getLogger(), "Reset Servo Solver");
+  joint_cmd_rolling_window_.clear();
+  return true;
+};
+
+moveit_msgs::action::LocalPlanner::Feedback
+ServoSolver::solve(const robot_trajectory::RobotTrajectory& local_trajectory,
+                   const std::shared_ptr<const moveit_msgs::action::LocalPlanner::Goal> /*local_goal*/,
+                   trajectory_msgs::msg::JointTrajectory& local_solution)
+{
+  // Create Feedback
+  moveit_msgs::action::LocalPlanner::Feedback feedback_result;
+
+  // Transform next robot trajectory waypoint into JointJog message
+  moveit_msgs::msg::RobotTrajectory robot_command;
+  local_trajectory.getRobotTrajectoryMsg(robot_command);
+
+  if (robot_command.joint_trajectory.points.empty())
+  {
+    feedback_result.feedback = std::string("Reference trajectory does not contain any points");
+    return feedback_result;
+  }
+
+  // Get current state
+  const auto current_state = planning_scene_monitor_->getStateMonitor()->getCurrentState();
+
+  // Create goal state
+  moveit::core::RobotState target_state(local_trajectory.getRobotModel());
+  target_state.setVariablePositions(robot_command.joint_trajectory.joint_names,
+                                    robot_command.joint_trajectory.points[0].positions);
+  target_state.update();
+
+  // TF planning_frame -> current EE
+  Eigen::Isometry3d current_pose = current_state->getFrameTransform(solver_parameters_.reference_frame);
+  // TF planning -> target EE
+  Eigen::Isometry3d target_pose = target_state.getFrameTransform(solver_parameters_.reference_frame);
+
+  // current EE -> planning frame * planning frame -> target EE
+  Eigen::Isometry3d diff_pose = current_pose.inverse() * target_pose;
+  Eigen::AngleAxisd axis_angle(diff_pose.linear());
+
+  constexpr double fixed_trans_vel = 0.05;
+  constexpr double fixed_rot_vel = 5;
+  const double trans_gain = fixed_trans_vel / diff_pose.translation().norm();
+  const double rot_gain = fixed_rot_vel / diff_pose.rotation().norm();
+
+  // Calculate Cartesian command delta
+  // Transform current pose to command frame
+  // Transform goal pose to command frame
+  servo_->setCommandType(moveit_servo::CommandType::TWIST);
+  moveit_servo::TwistCommand target_twist{
+    solver_parameters_.reference_frame,
+    { diff_pose.translation().x() * trans_gain, diff_pose.translation().y() * trans_gain,
+      diff_pose.translation().z() * trans_gain, axis_angle.axis().x() * axis_angle.angle() * rot_gain,
+      axis_angle.axis().y() * axis_angle.angle() * rot_gain, axis_angle.axis().z() * axis_angle.angle() * rot_gain }
+  };
+
+  std::optional<trajectory_msgs::msg::JointTrajectory> trajectory_msg;
+  // Create servo commands until a trajectory message can be generated
+  while (!trajectory_msg)
+  {
+    // Calculate next servo command
+    moveit_servo::KinematicState joint_state = servo_->getNextJointState(current_state, target_twist);
```

Comment:
> This point may cause issues if not handled with caution. Servo has a parameter called publish_period which is the time difference used for numerical integration. So if you ask servo to move at a velocity of 1 m/s and publish_period was .1 sec, then servo will return a new position that correspondence to a .1 m offset from the current position. So you should consider setting a faction of local trajectory length. For example, if you are supposed to arrive at your next point in .3 sec, then the update period should be .3/n, where n is the number of points you want in the command queue. I would recommend n=3.

Reply:
I am setting the (hopefully) correct publish_period in the initialize() function (see [diff](https://github.com/ros-planning/moveit2/pull/2673/commits/96160dab8d68bd565af4ceb19cbfc54e3b4f51cc)). Do you mind validating this?

Reply:
It looks correct. The only issue I have with it now is that `getNextJointState` keeps getting called until `trajectory_msg` is a value. if for some reason  `moveit_servo::composeTrajectoryMessage` is changed to only return once 4 values are in the trajectory, then this would half break and be hard to detect. All that to say, I think you should add a check on the trajectory message to make sure it has length 3.

Reply:
This is correct as long as this loop always gets called three times. I think you should replace the whole loop with a for loop of length 3 to make sure this is the case. Then check at the end to make sure the message is not empty.